library/core/src/slice/sort.rs

```diff
@@ -831,6 +831,15 @@ fn partition_at_index_loop<'a, T, F>(
 ) where
     F: FnMut(&T, &T) -> bool,
 {
+    // Limit the amount of iterations and fall back to heapsort, similarly to `slice::sort_unstable`.
+    // This lowers the worst case running time from O(n^2) to O(n log n).
+    // FIXME: Investigate whether it would be better to use something like Median of Medians
```

Comment:
Wikipedia says that median-of-median is somewhat slow, and not used normally, so I don't know if we'd want to start there, but maybe using it as fallback instead of heapsort.  But that's a discussion for another time.

Reply:
Yes, what I meant was to use median of medians as a fallback instead of heapsort to keep the fast average case of quickselect and still ensure O(n) worst case.

Reply:
Yeah I think I agree that fallback to median-of-medians would make sense.

And it sounds like it is used in practice in *some* contexts, see: https://en.wikipedia.org/wiki/Introselect